napari/_qt/widgets/qt_command_palette.py

```diff
@@ -0,0 +1,431 @@
+from __future__ import annotations
+
+import re
+from collections.abc import Iterator, Mapping
+from typing import TYPE_CHECKING, Any, cast
+
+from app_model.types import CommandRule, MenuItem
+from qtpy import QtCore, QtGui, QtWidgets as QtW
+from qtpy.QtCore import Qt, Signal
+
+from napari._app_model import get_app_model
+from napari._app_model.context._context import get_context
+
+if TYPE_CHECKING:
+    from napari._qt.qt_main_window import _QtMainWindow
+
+
+class QCommandPalette(QtW.QWidget):
+    """A Qt command palette widget."""
+
+    hidden = Signal()
+
+    def __init__(self, parent: QtW.QWidget | None = None):
+        super().__init__(parent)
+
+        self._line = QCommandLineEdit()
+        self._list = QCommandList()
+        _layout = QtW.QVBoxLayout(self)
+        _layout.addWidget(self._line)
+        _layout.addWidget(self._list)
+        self.setLayout(_layout)
+
+        self._line.textChanged.connect(self._on_text_changed)
+        self._list.commandClicked.connect(self._on_command_clicked)
+        self._line.editingFinished.connect(self.hide)
+        font = self.font()
+        font.setPointSize(16)
+        self.setFont(font)
+        self._line.setFont(font)
+        self.hide()
+
+        app = get_app_model()
+        # this appears to be a flat list of menu items, even though the
+        # type hint suggests menu or submenu
+        menu_items = app.menus.get_menu(app.menus.COMMAND_PALETTE_ID)
+        self.extend_command(
```

Comment:
The reason why @jni and I decided to put a separate command list in command palette is because the command palette has different order of commands. In the current situation, the recently executed commands will appear on the top, but in the future "my favorite commands" or "frequently used commands" will also rely on this.

> Thinking about it more, we could also just update the command list just before the user opens the command palette, instead of updating whenever there has been a change to the command registry (/menu change as you have done here). 

I noticed this only recently 😅 totally agree.